crypto/dh/dh_asn1.c

```diff
@@ -34,7 +34,11 @@ static int dh_cb(int operation, ASN1_VALUE **pval, const ASN1_ITEM *it,
         *pval = NULL;
         return 2;
     } else if (operation == ASN1_OP_D2I_POST) {
-        ((DH *)*pval)->dirty_cnt++;
+        DH *dh = (DH *)*pval;
+
+        DH_clear_flags(dh, DH_FLAG_TYPE_MASK);
```

Comment:
I assume this cannot be NULL?

Reply:
Does this mean for DHX that you now set both flags - and it did not before?

Reply:
> I assume this cannot be NULL?

The previously existing line assumed that, and so does the rest of this function, so...

Reply:
> Does this mean for DHX that you now set both flags - and it did not before?

Both `d2i_DHparams()` and `d2i_DHxparams()` failed to set their respective type flag.  That's fine as long as you *only* decode via the `dh_ameth.c` code, but direct calls would leave those bits zero, which is the `DH_FLAG_TYPE_DH` type flag.  Essentially, this code is mostly for good measure (and to avoid surprises if `DH_FLAG_TYPE_DH` would change to a non-zero value some time in the future), the added code in `d2i_DHxparams()` is the more crucial bit.